File: claimshield-ai/src/ai_reasoning.py
import os
import json

from google import genai


def explain_decision(
    claim,
    completeness,
    consistency,
    policy_results,
    decision_result
):
    api_key = os.getenv("GEMINI_API_KEY")

    if not api_key:
        return {
            "explanation": (
                "Gemini AI is unavailable because GEMINI_API_KEY is not configured. "
                "The deterministic decision engine result is retained."
            ),
            "aiAvailable": False,
        }

    try:
        client = genai.Client(api_key=api_key)

        evidence = {
            "claim": claim,
            "completeness": completeness,
            "consistency": consistency,
            "policyResults": policy_results,
            "deterministicDecision": decision_result,
        }

        prompt = f"""
You are an insurance claims evidence review assistant.

Your job is ONLY to explain the deterministic decision already produced
by the rules engine.

IMPORTANT RULES:
1. Do not change or override the deterministic decision.
2. Use ONLY the supplied claim evidence, consistency results,
   policy results, and deterministic decision.
3. Do not invent documents, facts, policy clauses, or evidence.
4. Mention the relevant policy number when available.
5. Clearly explain why the claim received its decision.
6. If evidence is missing, identify the specific missing evidence.
7. If information is inconsistent, clearly mention the inconsistency.
8. Keep the explanation concise and professional.
9. This is an explanation, not an independent claim decision.

Verified review data:

{json.dumps(evidence, indent=2)}

Return a short explanation suitable for an insurance claim reviewer.
"""

        response = client.models.generate_content(
            model="gemini-2.5-flash",
            contents=prompt
        )

        explanation = response.text.strip()

        if not explanation:
            raise ValueError("Gemini returned an empty explanation.")

        return {
            "explanation": explanation,
            "aiAvailable": True,
        }

    except Exception as error:
        return {
            "explanation": (
                "Gemini explanation could not be generated. "
                "The deterministic decision engine result is retained. "
                f"AI error: {str(error)}"
            ),
            "aiAvailable": False,
        }

File: claimshield-ai/server/lib/aiReasoning.js
// GROUNDED AI REASONING LAYER.
// This layer NEVER decides anything. It only produces a human-readable
// explanation grounded strictly in the deterministic outputs it is given
// (completeness, consistency, policy results, decision). If an
// ANTHROPIC_API_KEY is configured, it asks Claude to phrase the explanation
// naturally citing only the facts provided. If the key is missing or the
// call fails, it falls back to a deterministic template — the claim is
// still fully reviewable, just without the natural-language polish.

const MODEL = "claude-sonnet-4-6";

function buildGroundedPrompt({ claim, completeness, consistency, policyResults, decisionResult }) {
  return `You are an explainability assistant for an insurance claims review system.
You must ONLY restate and connect facts given below. Do not invent any evidence,
dates, amounts, or policy clauses that are not listed here. Keep the explanation
to 2-4 sentences, professional and neutral in tone.

Claim: ${claim.id} — ${claim.customer}, ${claim.vehicle}
Incident type: ${claim.incidentType}
Evidence available: ${completeness.available.map((a) => a.label).join(", ") || "none"}
Evidence missing: ${completeness.missing.map((m) => m.label).join(", ") || "none"}
Consistency: ${consistency.isConsistent ? "consistent" : "contradictions found - " + consistency.issues.map((i) => i.description).join(" ")}
Policy clause results: ${policyResults
    .map((p) => `Clause ${p.number} (${p.title}) - ${p.status}`)
    .join("; ") || "none matched"}
Final decision already made by the deterministic engine: ${decisionResult.decision} — ${decisionResult.reason}

Write a short explanation of why this decision follows from the facts above.`;
}

function fallbackExplanation({ decisionResult }) {
  return decisionResult.reason;
}

export async function explainDecision(context) {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    return { explanation: fallbackExplanation(context), aiAvailable: false };
  }

  try {
    const res = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "x-api-key": apiKey,
        "anthropic-version": "2023-06-01",
      },
      body: JSON.stringify({
        model: MODEL,
        max_tokens: 300,
        messages: [{ role: "user", content: buildGroundedPrompt(context) }],
      }),
    });

    if (!res.ok) throw new Error(`AI service returned ${res.status}`);
    const data = await res.json();
    const text = (data.content || [])
      .filter((b) => b.type === "text")
      .map((b) => b.text)
      .join(" ")
      .trim();

    if (!text) throw new Error("Empty AI response");
    return { explanation: text, aiAvailable: true };
  } catch (err) {
    // AI reasoning is optional polish — never let its failure block a decision.
    return { explanation: fallbackExplanation(context), aiAvailable: false };
  }
}
